src/Serialization/Rings.jl

```diff
@@ -88,161 +95,185 @@ end
-@registerSerializationType(fpPolyRingElem)
-@registerSerializationType(zzModPolyRingElem)
-
 function save_internal(s::SerializerState, p::PolyRingElem)
+    # store a polynomial over a ring provided we can store elements in that ring
     parent_ring = parent(p)
-    parent_ring = save_type_dispatch(s, parent_ring)
+    base = base_ring(parent_ring)
+    coeffs = coefficients(p)
+    encoded_terms = []
+    # flattened tree required for coefficients having parents
+    parents = []
+    exponent = 0
+    for coeff in coeffs
+        # collect only non trivial terms
+        if is_zero(coeff)
+            exponent += 1
+            continue
+        end
+        
+        encoded_coeff = save_internal(s, coeff)
+        if has_elem_basic_encoding(base)
+            push!(encoded_terms,  (exponent, encoded_coeff))
+        else
+            parents = encoded_coeff[:parents]
+            push!(encoded_terms,  (exponent, encoded_coeff[:terms]))
+        end
+        exponent += 1
+    end
+    parent_ring = save_as_ref(s, parent_ring)
+    # end of list should be loaded last
 
+    push!(parents, parent_ring)
     return Dict(
-        :parent => parent_ring,
-        :coeffs => save_type_dispatch(s, collect(coefficients(p)))
+        :parents => parents,
+        :terms =>  encoded_terms
     )
 end
 
-function load_internal(s::DeserializerState, ::Type{<: PolyRingElem}, dict::Dict)
-    R = load_unknown_type(s, dict[:parent])
-    coeff_ring = coefficient_ring(R)
-    coeff_type = elem_type(coeff_ring)
-    coeffs = load_type_dispatch(s, Vector{coeff_type}, dict[:coeffs])
+function load_terms(s::DeserializerState, parents::Vector, terms::Vector,
+                    parent_ring::PolyRing)
+    if isempty(terms)
+        return parent_ring(0)
+    end
+    
+    # shift so constant starts at 1
+    degree = max(map(x->x[1] + 1, terms)...)
+    base = base_ring(parent_ring)
+    loaded_terms = zeros(base, degree)
+
+    for term in terms
+        exponent, coeff = term
+        # account for shift
+        exponent += 1
+
+        if length(parents) == 1
+            @assert has_elem_basic_encoding(base)
+            coeff_type = elem_type(base)
+            loaded_terms[exponent] = load_internal_with_parent(s, coeff_type, coeff, base)
+        else
+            loaded_terms[exponent] = load_terms(s, parents[1:end - 1], coeff, parents[end - 1])
+        end
+    end
+    return parent_ring(loaded_terms)
+end
 
-    return R(coeffs)
+function load_terms(s::DeserializerState, parents::Vector, terms::Vector,
+                    parent_ring::Union{MPolyRing, UniversalPolyRing})
+    base = base_ring(parent_ring)
+    polynomial = MPolyBuildCtx(parent_ring)
+    for term in terms
+        e, coeff = term
+        if length(parents) == 1
+            coeff_type = elem_type(base)
+            if has_elem_basic_encoding(base)
+                c = load_type_dispatch(s, coeff_type, coeff, parent=base)
+            else
+                c = base(c)
```

Comment:
Maybe this?
```suggestion
                c = base(coeff)
```

Reply:
Or if you can't get here, maybe add an `@assert` or so?

Reply:
line 208 could never be reached